src/business-logic/services/rundown-lock-service.ts

```diff
@@ -0,0 +1,125 @@
+import { Part } from '../../model/entities/part'
+import { Piece } from '../../model/entities/piece'
+import { Owner } from '../../model/enums/owner'
+import { InTransition } from '../../model/value-objects/in-transition'
+import { RundownService } from './interfaces/rundown-service'
+
+const RUNDOWN_LOCK_INTERVAL = 500
+const RUNDOWN_LOCK_ERROR_TEXT = 'Unable to do action. An action was already executed less than 500ms ago'
+
+export class RundownLockService implements RundownService {
+  private static instance: RundownService
+  private hasEnoughTimePassed: boolean = true
+
+  constructor(private readonly rundownService: RundownService) {}
+
+  public static getInstance(rundownService: RundownService): RundownService {
+    if (!this.instance) {
+      this.instance = new RundownLockService(rundownService)
+    }
+    return this.instance
+  }
+
+  public takeNext(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
+    return this.rundownService.takeNext(rundownId)
+  }
+
+  public deleteRundown(rundownId: string): Promise<void> {
+    if (!this.hasEnoughTimePassed) {
+      throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
+    }
+    this.hasEnoughTimePassed = false
+    setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
```

Comment:
You are duplicating these lines in all methods that you want a lock for.
Whenever you start duplicating yourself, there should be something inside you telling you, that you are doing something suboptimally.

When you find yourself duplicating the same thing, that should be a clear indicator, that you atleast could extract it to a method.
```
private assertEnoughTimeHasPassed(): void {
  if (!this.hasEnoughTimePassed) {
    throw new Error(RUNDOWN_LOCK_ERROR_TEXT)
  }
  this.hasEnoughTimePassed = false
  setTimeout(() => (this.hasEnoughTimePassed = true), RUNDOWN_LOCK_INTERVAL)
}
```
The above snippet does the exact same thing as you implemented, but now you just need to call the method instead of rewriting it, everytime you need it.

Reply:
That logic is rework in previous comment